### PR TITLE
Engine - Implementation Sophos/Cfilter decoder

### PR DESCRIPTION
**Pull request by NahuFigueroa97**
|Related issue|
|---|
|#15851|

## Description

With the aim of expanding the coverage of the decoders, this PR adds a decoder for Sophos/Cfilter.

## Configuration options

Validate the new decoder:
- ./build/main catalog validate decoder/sophos-cfilter/0 < /home/vagrant/engine/wazuh/src/engine/ruleset/decoders/sophos/cfilter.yml

Create decoder:
- ./build/main  catalog create decoder < /home/vagrant/engine/wazuh/src/engine/ruleset/decoders/sophos/cfilter.yml

Update environment
- ./build/main catalog update environment/wazuh/0 < /home/vagrant/engine/wazuh/src/engine/ruleset/environments/wazuh-environment.yml

Test new decoder
- ./build/main test -q 1

## Logs/Alerts example

- <30>device="SFW" date=2020-05-18 time=14:38:53 timezone="CEST" device_name="XG230" device_id=1234567890123456 log_id=050901616001 log_type="Content Filtering" log_component="HTTP" log_subtype="Allowed" status="" priority=Information fw_rule_id=2 user_name="" user_gp="" iap=13 category="Information Technology" category_type="Acceptable" url="http://update.eset.com/eset_upd/ep7/dll/update.ver.signed" contenttype="" override_token="" httpresponsecode="" src_ip=1.128.3.4 dst_ip=91.228.167.133 protocol="TCP" src_port=65391 dst_port=80 sent_bytes=980 recv_bytes=295 domain=update.eset.com exceptions=av,https,sandstorm activityname="" reason="" user_agent="EFSW Update (Windows; U; 64bit; BPC 7.1.12010.0; OS: 10.0.17763 SP 0.0 NT; TDB 45511; CL 1.1.1; x64s; APP efsw; PX 1; PUA 1; CD 1; RA 1; PEV 0; UNS 1; UBR 1158; HVCI 0; SHA256 1; WU 3; HWF: 01009DAA-757A-D666-EFD2-92DD0D501284; PLOC de_de; PCODE 211.0.0; " status_code="304" transactionid="" referer="" download_file_name="" download_file_type="" upload_file_name="" upload_file_type="" con_id=248426360 application="" app_is_cloud=0 override_name="" override_authorizer=""

## Tests

Event:

- <30>device="SFW" date=2020-05-18 time=14:38:53 timezone="CEST" device_name="XG230" device_id=1234567890123456 log_id=050901616001 log_type="Content Filtering" log_component="HTTP" log_subtype="Allowed" status="" priority=Information fw_rule_id=2 user_name="" user_gp="" iap=13 category="Information Technology" category_type="Acceptable" url="http://update.eset.com/eset_upd/ep7/dll/update.ver.signed" contenttype="" override_token="" httpresponsecode="" src_ip=1.128.3.4 dst_ip=91.228.167.133 protocol="TCP" src_port=65391 dst_port=80 sent_bytes=980 recv_bytes=295 domain=update.eset.com exceptions=av,https,sandstorm activityname="" reason="" user_agent="EFSW Update (Windows; U; 64bit; BPC 7.1.12010.0; OS: 10.0.17763 SP 0.0 NT; TDB 45511; CL 1.1.1; x64s; APP efsw; PX 1; PUA 1; CD 1; RA 1; PEV 0; UNS 1; UBR 1158; HVCI 0; SHA256 1; WU 3; HWF: 01009DAA-757A-D666-EFD2-92DD0D501284; PLOC de_de; PCODE 211.0.0; " status_code="304" transactionid="" referer="" download_file_name="" download_file_type="" upload_file_name="" upload_file_type="" con_id=248426360 application="" app_is_cloud=0 override_name="" override_authorizer=""

Output:
```
{
    "wazuh": {
        "queue": 49,
        "origin": "/dev/stdin"
    },
    "event": {
        "original": "<30>device=\"SFW\" date=2020-05-18 time=14:38:53 timezone=\"CEST\" device_name=\"XG230\" device_id=1234567890123456 log_id=050901616001 log_type=\"Content Filtering\" log_component=\"HTTP\" log_subtype=\"Allowed\" status=\"\" priority=Information fw_rule_id=2 user_name=\"\" user_gp=\"\" iap=13 category=\"Information Technology\" category_type=\"Acceptable\" url=\"http://update.eset.com/eset_upd/ep7/dll/update.ver.signed\" contenttype=\"\" override_token=\"\" httpresponsecode=\"\" src_ip=1.128.3.4 dst_ip=91.228.167.133 protocol=\"TCP\" src_port=65391 dst_port=80 sent_bytes=980 recv_bytes=295 domain=update.eset.com exceptions=av,https,sandstorm activityname=\"\" reason=\"\" user_agent=\"EFSW Update (Windows; U; 64bit; BPC 7.1.12010.0; OS: 10.0.17763 SP 0.0 NT; TDB 45511; CL 1.1.1; x64s; APP efsw; PX 1; PUA 1; CD 1; RA 1; PEV 0; UNS 1; UBR 1158; HVCI 0; SHA256 1; WU 3; HWF: 01009DAA-757A-D666-EFD2-92DD0D501284; PLOC de_de; PCODE 211.0.0; \" status_code=\"304\" transactionid=\"\" referer=\"\" download_file_name=\"\" download_file_type=\"\" upload_file_name=\"\" upload_file_type=\"\" con_id=248426360 application=\"\" app_is_cloud=0 override_name=\"\" override_authorizer=\"\"",
        "action": "allowed",
        "code": 50901616001,
        "dataset": "sophos.xg",
        "end": "2020-05-18T14:38:53",
        "kind": "event",
        "module": "sophos",
        "outcome": "success",
        "severity": 6,
        "start": "2020-05-18T14:38:53",
        "timezone": "CEST"
    },
    "destination": {
        "bytes": 295,
        "ip": "91.228.167.133",
        "port": 80
    },
    "fileset": {
        "name": "xg"
    },
    "http": {
        "response": {
            "status_code": 304
        }
    },
    "input": {
        "type": "log"
    },
    "log": {
        "level": "Information"
    },
    "message": "tcp",
    "observer": {
        "product": "XG",
        "serial_number": 1234567890123456,
        "type": "firewall",
        "vendor": "Sophos"
    },
    "network": {
        "protocol": "http",
        "transport": "tcp"
    },
    "service": {
        "type": "sophos"
    },
    "sophos": {
        "xg": {
            "app_is_cloud": 0,
            "category": "Information Technology",
            "category_type": "Acceptable",
            "con_id": 248426360,
            "device": "SFW",
            "device_name": "XG230",
            "exceptions": "av,https,sandstorm",
            "fw_rule_id": 2,
            "iap": 13,
            "log_component": "HTTP",
            "log_id": 50901616001,
            "log_subtype": "Allowed",
            "log_type": "Content Filtering",
            "priority": "Information",
            "status_code": 304
        }
    },
    "source": {
        "bytes": 980,
        "ip": "1.128.3.4",
        "port": 65391
    },
    "tags": [
        "forwarded",
        "preserve_original_event",
        "sophos-xg"
    ],
    "\\@timestamp": "2020-05-18T14:38:53",
    "url": {
        "domain": "update.eset.com",
        "full": "http://update.eset.com/eset_upd/ep7/dll/update.ver.signed",
        "original": "http://update.eset.com/eset_upd/ep7/dll/update.ver.signed",
        "path": "/eset_upd/ep7/dll/update.ver.signed",
        "scheme": "http",
        "extension": "signed"
    },
    "related": {
        "ip": [
            "91.228.167.133",
            "1.128.3.4"
        ]
    },
    "user_agent": {
        "original": "EFSW Update (Windows; U; 64bit; BPC 7.1.12010.0; OS: 10.0.17763 SP 0.0 NT; TDB 45511; CL 1.1.1; x64s; APP efsw; PX 1; PUA 1; CD 1; RA 1; PEV 0; UNS 1; UBR 1158; HVCI 0; SHA256 1; WU 3; HWF: 01009DAA-757A-D666-EFD2-92DD0D501284; PLOC de_de; PCODE 211.0.0; ",
        "os": {
            "name": "Windows"
        }
    }
}
```

### CODE REVIEW

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    map:
+      - event.severity: 4
+  - check:
+      - ~tmp.priority: Notification
+    map:
+      - event.severity: 5
+  - check:
+      - ~tmp.priority: Information
+    map:
+      - event.severity: 6
```
Comment (JorgeMarinoDev) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:72]: Move event severity mapping to the end in a new section with proper comment.

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    map:
+      - ~tmp.aux_slash: /
+      - url.full: $~tmp.url
+      - url.original: $~tmp.url
+      - url.path: +s_concat/$~tmp.aux_slash/$~tmp.path
+      - url.scheme: $~tmp.schema
+  - logpar:
+      - url.path: /<~tmp.fill>/<~tmp.fill>/<~tmp.fill>/<~tmp.fill>.<~tmp.fill>.<~tmp.extension>
+    map:
+      - url.extension: $~tmp.extension
```
Comment (JorgeMarinoDev) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:121]: Same discussion about url parsing.
https://github.com/wazuh/wazuh/pull/16063#discussion_r1093496001

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  - map:
+      - related.ip: [$destination.ip, $source.ip]
+      - user_agent.original: $~tmp.user_agent
```
Comment (JorgeMarinoDev) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:124]: Same comment as here.
https://github.com/wazuh/wazuh/pull/16063#discussion_r1093502976

**Review by JorgeMarinoDev — CHANGES_REQUESTED** [review]
Looks good to me. A few readability improvements can be made and check the url parsing section.

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
+---
+name: decoder/sophos-cfilter/0
+
+metadata:
+  module: sophos
+  title: Sophos logs decoder
+  description: Decoder for sophos logs
```
Comment (soynof) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:7]: Improve (Documentation): add compatibility version

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    - event.original: \<<~log.head_message>><~log.payload_message>
```
Comment (soynof) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:24]: Improve (Design): if is possible use only one temporal variable, like ~tmp

File: src/engine/ruleset/decoders/sophos/cfilter.yml [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+normalize:
+  - check:
```
Comment (soynof) [inline src/engine/ruleset/decoders/sophos/cfilter.yml:27]: Improve (Design): check is not necessary

**Pull request merged**